### PR TITLE
Facilitate call to DetectorClocksStandard::DataFor() [2/2]

### PR DESCRIPTION
**Pull request by PetrilloAtWork**
This is part of a group of pull requests that also include pull request LArSoft/lardataalg#16.

The explanation and details of this request are stated in that pull request.
Please keep the discussion on the general design of the feature in there.

### CODE REVIEW

**Comment by knoepfel** [comment]
This change does not need to wait for L2 approval, assuming successful CI tests.

**Comment by lgarren** [comment]
trigger build with pull request #11 lardataalg#16

**Comment by lgarren** [comment]
trigger build with pull request #11 LArSoft/lardataalg#16

**Comment by lgarren** [comment]
Had to trigger the build manually.  @PetrilloAtWork would you take a look?
FAILED:
https://dbweb0.fnal.gov/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5644&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/1232&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/5617&builds=
OK:
https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9819&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:ArgoNeuT/view_builds/index?offset=0&builds=argoneut_ci/5239&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:DUNE/view_builds/index?offset=0&builds=dune_ci/10414&builds=

**Comment by PetrilloAtWork** [comment]
**TL;DR** Fix to `ubana` is in `feature/gp_simplerForData`, expert review appreciated.

Thank you for the feedback.
The ICARUS and SBND errors apparently stem from an unrelated error in `sbncode` (which has recently become their common dependency).
MicroBooNE's is definitely related to my change.
I ask some SciSoft expert (@knoepfel?) to review the following approach.
The [error](https://dbweb0.fnal.gov/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5644&builds=):
```2586: ubana/ubana/Calibrations/CMakeFiles/ExtractPedestalInformation_module.dir/ExtractPedestalInformation_module.cc.o: In function `__static_initialization_and_destruction_0':
2587: /cvmfs/larsoft.opensciencegrid.org/products/art/v3_05_01/include/art/Utilities/SharedResource.h:15: undefined reference to `typeinfo for detinfo::DetectorClocksServiceStandard'```
involves the linking of `detinfo::DetectorClocksServiceStandard`, which is a service interface implementation, into the translation unit of MicroBooNE module `ExtractPedestalInformation`.
Service implementations are not meant, in general, to be invoked directly, but rather via their interface. In that case, no linkage to the service implementation is required.
**But** if the code actually really needs a specific implementation, then it should also explicitly link to it.
MicroBooNE module:
1. does include `lardata/DetectorInfoServices/DetectorClocksServiceStandard.h` header directly
2. does _not_ use `detinfo::DetectorClocksServiceStandard`
    1. _(in fact, it does not even use `detinfo::DetectorClocksService` interface)_
3. does _not_ explicitly link to `detinfo_DetectorClocksServiceStandard_service` library

I guess that the _art_ service macro `DECLARE_ART_SERVICE_INTERFACE_IMPL` requires the implementation of the declared service to be available (or more precisely, it requires RTTI --- `typeinfo` --- and the virtual table of that service implementation to be available). GCC emits the virtual table since all the virtual functions are defined ([more precise explanation here](https://gcc.gnu.org/onlinedocs/gcc-8.2.0/gcc/Vague-Linkage.html#Vague-Linkage)) and all is good.

Then Gianluca comes, and decides that the implementation of one of the virtual functions of the service is pulling too many dependencies into the service header and it is better hidden, and he moves it to the implementation file of the service.
Then GCC finds that there is a virtual function without implementation (the new change) and that the implementation is not available in any of the linked libraries (point 3 above), so it does not emit the virtual table of the service class any more, and therefore its `typeinfo` is not available anymore, and the _art_ macro becomes unhappy because it really, really wanted that `typeinfo`.

The conclusion from my premise and analysis is that MicroBooNE module should not include the service (point 1), considering point 2 (and point 2.i) above.
So my proposed solution is to change MicroBooNE code, which I have pushed as `feature/gp_simplerForData` branch of `ubana`.

By the way: according to the surgeon general, the use of `typeid()` needs to be approved by an existing `#include <typeinfo>`, which `art/Utilities/SharedResource.h` appears to lack. I was hoping to blame _art_ for the issue, but I didn't get lucky.

**Comment by knoepfel** [comment]
@PetrilloAtWork, your fix is the right approach: nobody should ever explicitly include `DetectorClocksServiceStandard.h`.  I will make sure to add the `<typeinfo>` header to `art/Utilities/SharedResource.h`.

**Comment by lgarren** [comment]
Sigh, this is going to take a while to sort out.  There is still a problem with the uboonecode build.  This time in ubcv. https://dbweb0.fnal.gov/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5647&builds=

@PetrilloAtWork  would you please "mrb g uboone_suite" to test the full MicroBooNE suite.  Thanks!

**Comment by PetrilloAtWork** [comment]
I have pushed a `feature/gp_simplerForData` into `ubcv` too: everybody is entitled to a `feature/gp_simplerForData` branch.
But it turns out no other repository in MicroBooNE software suite requires it.

----

Summary to date: branch `feature/gp_simplerForData` in
* `lardataalg`
* `lardata`
* `ubana`
* `ubcv`

**Comment by lgarren** [comment]
trigger build with pull requests #11 LArSoft/lardataalg#16 ubana@feature/gp_simplerForData ubcv@feature/gp_simplerForData

**Comment by lgarren** [comment]
CI results are normal, but the uboone suite has not yet been updated 
https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9864&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:DUNE/view_builds/index?offset=0&builds=dune_ci/10440&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/1254&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:SBND/view_builds/index?offset=0&builds=sbnd_ci/5634&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:ArgoNeuT/view_builds/index?offset=0&builds=argoneut_ci/5252&builds=
https://dbweb0.fnal.gov/LarCI/app/ns:uBooNE/view_builds/index?offset=0&builds=uboone_ci/5657&builds=

**Comment by lgarren** [comment]
uboone unit tests are good

**Comment by lgarren** [comment]
approve
